resources/user.go

```diff
@@ -0,0 +1,306 @@
+	HomeDir           *string `yaml:"homedir"`
+	AllowDuplicateUID bool    `yaml:"allowduplicateuid"`
+
+	recWatcher *recwatch.RecWatcher
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *UserRes) Default() Res {
+	return &UserRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *UserRes) Validate() error {
+	if obj.State != "exists" && obj.State != "absent" {
+		return fmt.Errorf("State must be 'exists' or 'absent'")
+	}
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *UserRes) Init() error {
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *UserRes) Watch() error {
+	var err error
+	obj.recWatcher, err = recwatch.NewRecWatcher(passwdFile, false)
+	if err != nil {
+		return err
+	}
+	defer obj.recWatcher.Close()
+
+	// notify engine that we're running
+	if err := obj.Running(); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var send = false // send event?
+	var exit *error
+
+	for {
+		if obj.debug {
+			log.Printf("Watching: %s", passwdFile) // attempting to watch...
+		}
+
+		select {
+		case event, ok := <-obj.recWatcher.Events():
+			if !ok { // channel shutdown
+				return nil
+			}
+			if err := event.Error; err != nil {
+				return errwrap.Wrapf(err, "Unknown %s watcher error", obj)
+			}
+			if obj.debug { // don't access event.Body if event.Error isn't nil
+				log.Printf("%s: Event(%s): %v", obj, event.Body.Name, event.Body.Op)
+			}
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event := <-obj.Events():
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+			//obj.StateOK(false) // dirty // these events don't invalidate state
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// CheckApply method for User resource.
+func (obj *UserRes) CheckApply(apply bool) (checkOK bool, err error) {
+	log.Printf("%s: CheckApply(%t)", obj, apply)
+
+	var exists = true
+	usr, err := user.Lookup(obj.GetName())
+	if err != nil {
+		if _, ok := err.(user.UnknownUserError); !ok {
+			return false, errwrap.Wrapf(err, "error looking up user")
+		}
+		log.Printf("the user: %s does not exist", obj.GetName())
+		exists = false
+	}
+
+	if obj.AllowDuplicateUID == false && obj.UID != nil {
+		existingUID, err := user.LookupId(strconv.Itoa(int(*obj.UID)))
+		if err != nil {
+			if _, ok := err.(user.UnknownUserIdError); !ok {
+				return false, errwrap.Wrapf(err, "error looking up UID")
+			}
+		} else if existingUID.Username != obj.GetName() {
+			return false, errwrap.Wrapf(err, "the requested UID is already taken")
+		}
+	}
+
+	if usercheck := true; exists && obj.State == "exists" {
+		if obj.GID != nil && strconv.Itoa(int(*obj.GID)) != usr.Gid {
```

Comment:
two small comments, which you don't *have* to fix for this to *work*, but if you're motivated you probably should. They're good learning examples of "even though it compiles and passes tests, it should probably be different".
i mention this as a positive mentoring tip, but only read further if you promise to remember:

1) in the struct and the code we follow the order: `uid`, `gid`, `homedir`... as a result, it's usually classy to be consistent where it makes sense to be. This is one case, except you did gid, and then uid.

2) After reading up on the docs, in turns out the usr.Uid is a string, since that's the common type for cross platform compat. While you're comparison is correct, and will probably work, you converted our int to a string and then compared the two. It's probably more correct to convert the string to and int, and then compare the ints.

Why is that? Well if the library returned an empty string, and we had a 400, then they'd not match, but we'd silently be eating the error, where as if you perform a string->int you'll have another error to check if it can't be converted, and you'll catch that lib error and return "can't convert uid to uint32" or whatever...

Make sense? If not, ping before patching, and make sure you get the error handling perfect :) I believe in you!

Reply:
I promise to remember. :)